### PR TITLE
[15.0][IMP] account_commission: link settlements and vendor bill that is created from settlements

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
After installing this module, there will be a settlement link button in vendor bill that is created for settlement in the form view and tree view like below image.
Screenshot from 2022-12-07 17-27-12

Screenshot from 2022-12-07 17-27-23

### CODE REVIEW

**Review by yostashiro — COMMENTED** [review]
@AungKoKoLin1997 Can you please add tests and also share images of the adjusted views.

File: account_commission/models/account_move.py [hunk 1]
```diff
@@ -21,6 +21,41 @@ class AccountMove(models.Model):
...
+            form_view = [(self.env.ref("commission.view_settlement_form").id, "form")]
+            if "views" in action:
+                action["views"] = form_view + [
+                    (state, view) for state, view in action["views"] if view != "form"
+                ]
+            else:
+                action["views"] = form_view
+            action["res_id"] = settlements.id
+        else:
+            action = {"type": "ir.actions.act_window_close"}
```
Comment (pedrobaeza) [inline account_commission/models/account_move.py]: No need of this. Simply don't populate any action in this case, but avoid this `else` that is not covered in tests.

File: account_commission/models/account_move.py [hunk 2]
```diff
@@ -21,6 +21,41 @@ class AccountMove(models.Model):
...
+
```
Comment (pedrobaeza) [inline account_commission/models/account_move.py]: Don't add empty lines in the middle of a method.

**Review by pedrobaeza — COMMENTED** [review]
What about the inverse one: settlement > invoice? Maybe a note logged on the settlement saying that the invoice X has been generated with the possibility to navigate from it?

**Comment by AungKoKoLin1997** [comment]
> What about the inverse one: settlement > invoice? Maybe a note logged on the settlement saying that the invoice X has been generated with the possibility to navigate from it?

@pedrobaeza For the inverse one, I think we have invoice field and there is no problem for lacking button in form.

Screen Shot 2022-12-07 at 21 17 26

**Comment by pedrobaeza** [comment]
OK, agreed. What you can change is the label to "Generated invoice".

**Comment by AungKoKoLin1997** [comment]
> OK, agreed. What you can change is the label to "Generated invoice".

Ok.  I will change label

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Please change the commit message to `[IMP] account_commission: Add link invoice > settlement`

**Review by pedrobaeza — APPROVED** [review]

File: account_commission/models/account_move.py [hunk 3]
```diff
@@ -21,6 +21,33 @@ class AccountMove(models.Model):
         compute="_compute_agents",
         search="_search_agents",
     )
+    settlement_count = fields.Integer(compute="_compute_settlement")
+    settlement_ids = fields.Many2many(
```
Comment (yostashiro) [inline account_commission/models/account_move.py]: Should this field technically be one2many given that there should be only one valid vendor bill for a settlement?

We may take this chance to update _compute_invoice_id() of commission.settlement to exclude cancelled vendor bills as well.
https://github.com/oca/commission/blob/b6715536aa7c115b9283d07e10f7ffae51268ee2/account_commission/models/commission_settlement.py#L44-L47

@pedrobaeza Any thoughts on this?

Comment (pedrobaeza) [reply account_commission/models/account_move.py]: OK for me.

File: account_commission/views/account_move_views.xml [hunk 4]
```diff
@@ -111,4 +126,22 @@
             </field>
         </field>
     </record>
+    <record id="view_in_invoice_bill_tree_inherit" model="ir.ui.view">
```
Comment (yostashiro) [inline account_commission/views/account_move_views.xml:119]: Please move this record after `view_move_form` record.

File: account_commission/models/commission_settlement.py [hunk 5]
```diff
@@ -30,6 +30,7 @@ class CommissionSettlement(models.Model):
         readonly=True,
     )
     invoice_id = fields.Many2one(
+        string="Generated invoice",
```
Comment (yostashiro) [inline account_commission/models/commission_settlement.py]:
```suggestion
        string="Generated Invoice",
```

File: account_commission/models/commission_settlement.py [hunk 6]
```diff
@@ -30,6 +30,7 @@ class CommissionSettlement(models.Model):
...
         store=True,
         comodel_name="account.move",
         compute="_compute_invoice_id",
```
Comment (yostashiro) [inline account_commission/models/commission_settlement.py:36]: Please update the content of the compute method to:
```python
        for record in self:
            record.invoice_id = record.invoice_line_ids.filtered(lambda x: x.parent_state != "cancel")[:1].move_id
```

File: account_commission/models/account_move.py [hunk 7]
```diff
@@ -21,6 +21,33 @@ class AccountMove(models.Model):
...
+        "commission.settlement",
+        string="Settlements",
+        compute="_compute_settlement",
+        copy=False,
+    )
```
Comment (yostashiro) [inline account_commission/models/account_move.py]:
```suggestion
    settlement_ids = fields.One2many(
        "commission.settlement",
        string="Settlements",
        compute="_compute_settlement",
    )
```

File: account_commission/models/account_move.py [hunk 8]
```diff
@@ -21,6 +21,33 @@ class AccountMove(models.Model):
...
+        settlements = self.mapped("settlement_ids")
+        if not settlements or len(settlements) > 1:
+            action["domain"] = [("id", "in", settlements.ids)]
+        elif len(settlements) == 1:
+            res = self.env.ref("commission.view_settlement_form", False)
+            action["views"] = [(res and res.id or False, "form")]
+            action["res_id"] = settlements.id
+        return action
+
+    @api.depends("invoice_line_ids")
```
Comment (yostashiro) [inline account_commission/models/account_move.py]:
```suggestion
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
Mostly nitpicking.

**Review by yostashiro — APPROVED** [review]
Code review. 👍

**Comment by pedrobaeza** [comment]
Please squash commits into one.

**Comment by AungKoKoLin1997** [comment]
> Please squash commits into one.

Done!

**Pull request merged**